eval/topdown.go

```diff
@@ -540,6 +588,174 @@ func evalRefRec(ctx *TopDownContext, ref opalog.Ref, iter TopDownIterator, path
 	}
 }
 
+func evalRefRule(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+
+	switch rule.DocKind() {
+	case opalog.PartialSetDoc:
+		return evalRefRulePartialSetDoc(ctx, ref, path, rule, iter)
+	case opalog.PartialObjectDoc:
+		return evalRefRulePartialObjectDoc(ctx, ref, path, rule, iter)
+	case opalog.CompleteDoc:
+		return evalRefRuleCompleteDoc(ctx, ref, path, rule, iter)
+	default:
+		panic(fmt.Sprintf("illegal argument: %v", rule))
+	}
+
+}
+
+func evalRefRuleCompleteDoc(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+	suffix := ref[len(path):]
+	if len(suffix) == 0 {
+		return fmt.Errorf("not implemented: %v %v %v", ref, path, rule)
+	}
+
+	bindings := newHashMap()
+	child := ctx.Child(rule, bindings)
+
+	return TopDown(child, func(child *TopDownContext) error {
+		switch v := rule.Value.Value.(type) {
+		case opalog.Object:
+			return evalRefRuleResult(ctx, ref, suffix, v, iter)
+		case opalog.Array:
+			return evalRefRuleResult(ctx, ref, suffix, v, iter)
+		default:
+			return fmt.Errorf("cannot dereference value (%T) in %v", rule.Value.Value, rule)
+		}
+	})
+}
+
+func evalRefRulePartialObjectDoc(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+	suffix := ref[len(path):]
+	if len(suffix) == 0 {
+		return fmt.Errorf("not implemented: %v %v %v", ref, path, rule)
+	}
+
+	if !suffix[0].IsGround() {
```

Comment:
There is a bug here. We need to take the current bindings into account when preparing the child context. I've fixed this in a recent commit that will be part of the next PR. The same goes for the partial set case below.